internal/graph/track_check_resolver.go

```diff
@@ -0,0 +1,211 @@
+package graph
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+	"go.uber.org/zap"
+
+	"github.com/openfga/openfga/pkg/logger"
+	"github.com/openfga/openfga/pkg/tuple"
+)
+
+const (
+	trackerInterval = time.Duration(60) * time.Second
+)
+
+// TrackerCheckResolverOpt defines an option pattern that can be used to change the behavior of TrackerCheckResolver.
+type TrackerCheckResolverOpt func(checkResolver *TrackerCheckResolver)
+type resolutionTree struct {
+	tm   time.Time
+	hits *atomic.Uint64
+}
+
+type TrackerCheckResolver struct {
+	delegate CheckResolver
+	ticker   *time.Ticker
+	logger   logger.Logger
+	ctx      context.Context
+	nodes    sync.Map
+}
+
+var _ CheckResolver = (*TrackerCheckResolver)(nil)
+
+func WithTrackerLogger(logger logger.Logger) TrackerCheckResolverOpt {
+	return func(t *TrackerCheckResolver) {
+		t.logger = logger
+	}
+}
+
+func WithTrackerContext(ctx context.Context) TrackerCheckResolverOpt {
+	return func(t *TrackerCheckResolver) {
+		t.ctx = ctx
+	}
+}
+
+// Expired check the current tuple entry expiration.
+func (r *resolutionTree) expired() bool {
+	return time.Since(r.tm) > trackerInterval
+}
+
+// NewTrackCheckResolver creates an instance tracker Resolver.
+func NewTrackCheckResolver(opts ...TrackerCheckResolverOpt) *TrackerCheckResolver {
+	t := &TrackerCheckResolver{
+		ticker: time.NewTicker(trackerInterval),
+	}
+
+	for _, opt := range opts {
+		opt(t)
+	}
+
+	t.delegate = t
+	t.launchFlush()
+	return t
+}
+
+// LogExecutionPaths reports the model and tuple path.
+func (t *TrackerCheckResolver) logExecutionPaths() {
+	t.nodes.Range(func(k, v any) bool {
+		modelid := k.(string)
+		paths, _ := v.(*sync.Map)
+		paths.Range(func(k, v any) bool {
+			tree, _ := v.(*resolutionTree)
+			path := k.(string)
+			if tree.expired() {
+				t.logger.Info("execution path hits",
+					zap.String("modelid", modelid),
+					zap.String("path", path),
+					zap.Uint64("hits", tree.hits.Load()))
+				paths.Delete(path)
+			}
+			return true
+		})
+		return true
+	})
+}
+
+// LaunchFlush starts the execution path logging and removal of entries.
+func (t *TrackerCheckResolver) launchFlush() {
+	go func() {
+		for {
+			select {
+			case <-t.ctx.Done():
+				t.ticker.Stop()
+				return
+			case <-t.ticker.C:
+				t.logExecutionPaths()
+				t.ticker.Reset(trackerInterval)
```

Comment:
But this flushes all hits within the 60 seconds window rather than the batches of a storeId/modelId. This would result at at every 60 seconds we get a stampede of logs rather than small batches every other second as they get spread out.

Reply:
Every 60 seconds it reports path hits and removes only those paths. So, the logs would reflect only those paths that have not been reported